File: Top 100 Codes/05-Operations on Strings/20-Count common sub-sequence in two strings/Countcommonsubsequenceintwostrings.py
def longest_common_subsequence(str1, str2):
    m, n = len(str1), len(str2)

    # Create a 2D DP table initialized with 0
    dp = [[0] * (n + 1) for _ in range(m + 1)]

    # Fill the DP table
    for i in range(1, m + 1):
        for j in range(1, n + 1):
            if str1[i - 1] == str2[j - 1]:
                dp[i][j] = 1 + dp[i - 1][j - 1]  # Characters match
            else:
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])  # Take the max

    return dp[m][n]  # LCS length


# Input
str1 = input("Enter first string: ")
str2 = input("Enter second string: ")

# Output
lcs_length = longest_common_subsequence(str1, str2)
print("Length of Longest Common Subsequence:", lcs_length)

File: Top 100 Codes/05-Operations on Strings/20-Count common sub-sequence in two strings/Countcommonsubsequenceintwostrings.cpp
#include <iostream>
#include <vector>
#include <string>
using namespace std;

int longestCommonSubsequence(string A, string B) {
    int m = A.length();
    int n = B.length();

    // Create DP table
    vector<vector<int>> dp(m + 1, vector<int>(n + 1, 0));

    // Fill the table
    for (int i = 1; i <= m; i++) {
        for (int j = 1; j <= n; j++) {
            if (A[i - 1] == B[j - 1]) {
                dp[i][j] = 1 + dp[i - 1][j - 1];  // Match found
            } else {
                dp[i][j] = max(dp[i - 1][j], dp[i][j - 1]);  // Skip one character
            }
        }
    }

    return dp[m][n];  // LCS length
}

int main() {
    string str1, str2;
    cout << "Enter first string: ";
    cin >> str1;
    cout << "Enter second string: ";
    cin >> str2;

    int lcsLength = longestCommonSubsequence(str1, str2);
    cout << "Length of Longest Common Subsequence: " << lcsLength << endl;

    return 0;
}
